File: Algo/Leetcode/034FirstAndLastPosInSortedArray.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.size()==0) return vector<int>{-1, -1};

        int mid = findMid(nums, target);
        if(mid != -1){
            int l = mid;
            int r = mid;
            while(l>=0 && nums[l] == target) l--;
            while(r<nums.size() && nums[r] == target) r++;
            return vector<int>{l+1, r-1};
        }
        return vector<int>{-1, -1};

    }

    int findMid(vector<int>& nums, int target){
        int l = 0; int r = nums.size()-1;
        while(l<=r){
            int mid = (l+r) / 2;
            if(nums[mid] == target) return mid;
            else if(nums[mid] > target) r = mid-1;
            else l = mid+1;
        }
        return -1;
    }
};

File: Algo/Leetcode/034FirstAndLastPosInSortedArray.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        l = 0
        r = len(nums) - 1
        mid = self.findMid(l, r, nums, target)
        if mid != -1:
            l = mid
            r = mid
            while l >= 0 and nums[l] == target:
                l -= 1
            while r < len(nums) and nums[r] == target:
                r += 1
            return [l + 1, r - 1]

        return [-1, -1]


    def findMid(self, l, r, nums, target):
        """Binary search"""
        while l <= r:
            mid = (l + r) / 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                r = mid - 1
            else:
                l = mid + 1

        return -1